src/mono/mono/metadata/loader.c

```diff
@@ -81,30 +81,31 @@ static gint32 signatures_size;
 void
 mono_loader_init ()
 {
-	static gboolean inited;
+	static volatile gboolean inited = FALSE; // volatile to avoid race conditions
```

Comment:
`volatile` in C has nothing to do with atomicity or ordering. It means two things:

* Don't optimize away seemingly redundant loads/stores involving this location.
* Don't reorder loads/stores involving this location with respect to other `volatile` loads/stores.

No more, no less.

This code is still prone to race conditions. The correct fix would probably be to use `mono_atomic_cas`.

(Also, the initializer is unnecessary. Global variables in C are always zero-initialized.)

Reply:
In my experience, this has solved many hard to find bugs.

Also volatile ensures the value is in memory, so other threads reading it will be affected if the value is changed. That’s why the qualifier exists, in case it is modified by other processes or externalities.

Reply:
> In my experience, this has solved many hard to find bugs.

Sprinkling `volatile` on code until it works is not actually how you arrive at correct code. It just happens to work *for now*. In addition, in this particular case, I'm 100% certain you can still exercise a race condition in practice on some architectures.

There are absolutely situations where `volatile` is appropriate, but this isn't it.

> Also volatile ensures the value is in memory, so other threads reading it will be affected if the value is changed.

This is a vague statement that doesn't really mean anything. It's a global variable; it's in memory by definition.

You should think of `volatile` as nothing more than a mechanism to enforce ordering upon the C compiler's optimizer. What matters here is the run-time cross-thread ordering of memory operations, and `volatile` does absolutely nothing to help with that. With your change, it's still possible for 2 threads to see `inited` as `FALSE` and proceed with initialization. CAS is the appropriate construct here.

Reply:
> volatile in C has nothing to do with atomicity or ordering. It means two things:

But then you go on to say it does have to with ordering:

> Don't reorder loads/stores involving this location with respect to other volatile loads/stores

Which is indeed useful, assuming there are other relevant volatiles.
Volatile really was the semantically correct and cheap "atomic", for x86, until ARM came along (and PowerPC, Sparc, etc...and worst of all, Alpha).
And then, it does work on ARM, but is no longer cheap and I guess there are cheaper alternatives?
Esp. "half volatile", i.e. half barriers instead of full barriers.

Usually a better solution, ignoring Alpha, is to have "data dependencies" where the would-be-volatile variable, is a pointer that points to on-demand initialized data, which is highly similar to the old pattern of InterlockedCompareExchangPointer on a pointer that points to the data.

And more so, C++11 static locals with initializers are really quite nice in the vast majority of runtime environments (the Windiws implementation is kinda horrible, if you have to run on XP or in kernel, but for the most part, ok).

There is some question, in general, with on-demand initialization, as to run-exactly-once or run-at-least-once or run-maybe-multiple-times-concurrently-but-commit-once.

Reply:
> But then you go on to say it does have to with ordering:

Sure, I could have been more precise in my wording: It has nothing to do with *cross-thread* ordering.

> Volatile really was the semantically correct and cheap "atomic", for x86, until ARM came along (and PowerPC, Sparc, etc...and worst of all, Alpha).

It's really not, though. MSVC gives `volatile` acquire/release semantics on x86 and x64, but that's *bizarre* to say the least. No other compiler in widespread use that I'm aware of does this; they all follow the GCC/Clang definition which is much closer to the intent of the C standard.

For portability and sanity's sake, you really should be compiling with `/volatile:iso` if you're using MSVC so you don't end up relying on this historical quirk on x86/x64.